pfcpiface/p4rtc.go

```diff
@@ -49,6 +52,50 @@ type P4rtClient struct {
 	P4Info p4ConfigV1.P4Info
 }
 
+type P4RuntimeError struct {
+	errors []*p4.Error
+}
+
+func (e *P4RuntimeError) Error() string {
+	return fmt.Sprintf("P4RuntimeError: %v", e.errors)
+}
+
+func (e *P4RuntimeError) Get() []*p4.Error {
+	return e.errors
+}
+
+// convertError parses nested P4Runtime errors.
+// See https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-error-reporting-messages.
+func convertError(err error) error {
+	st, ok := status.FromError(err)
+	if !ok {
+		return err
+	}
+
+	if st.Code() != codes.Unknown {
+		return err
+	}
+
+	p4RtError := &P4RuntimeError{
+		errors: make([]*p4.Error, 0),
+	}
+
+	for _, detailItem := range st.Details() {
+		p4Error, ok := detailItem.(*p4.Error)
+		if !ok {
+			continue
+		}
+
+		if p4Error.GetCanonicalCode() == int32(codes.OK) {
+			continue
+		}
```

Comment:
The P4RT spec for batched RPCs says there must be one error message for each entity in the request, even if `ok`. I know that some code (not here, other projects) depend on the order and position of an individual error in the array to match an entity to its error. Discarding ok errors make it impossible to correlated _any_ error.

Pseudo code to illustrate

```python
request = [insert flow A, insert flow B]
response = [ok, fail]

for entity, idx in request:
  err = response[idx]
  if err == ALREADY_EXISTS:
    // discard
  if err == INVALID_PARAM:
    // handle
```

Reply:
Good point, the last commit should address that